torch/autograd/function.py

```diff
@@ -84,7 +84,15 @@ def _register_hook(backward_hooks, hook):
 class BackwardCFunction(_C._FunctionBase, _ContextMethodMixin, _HookMixin):
     def apply(self, *args):
         # _forward_cls is defined by derived class
-        return self._forward_cls.backward(self, *args)  # type: ignore[attr-defined]
+        # The user should define either backward or vjp but never both.
+        backward_fn = self._forward_cls.backward
+        vjp_fn = self._forward_cls.vjp
+        if backward_fn is not Function.backward and vjp_fn is not Function.vjp:
+            raise RuntimeError("Implementing both 'backward' and 'vjp' for a custom "
+                               "Function is not allowed. You should only implement one "
+                               "of them.")
+        user_fn = vjp_fn if vjp_fn is not Function.vjp else backward_fn
+        return user_fn(self, *args)  # type: ignore[attr-defined]
```

Comment:
Do we still need this `type: ignore[attr-defined]`?

Reply:
Given the lint failures, we do, but a few lines above. Let me fix that.